Added `big.mark` argument to incorporate thousands separator

kenkomodo opened the pull request:
Added `big.mark` argument to `as_fc()`, `fc_split()` and `fc_filter()` with defaults set to `","` so that comma is added to `N` or `n` for values >999. Including the argument allows flexibility for users that prefer an alternative thousands separator to change it accordingly.

Closes #26 

Exact same example code from issue:
```
as_fc(N = 3000) |> 
  fc_filter(
    N = 2400, 
    label = "Randomized patients", 
    show_exc = TRUE,
    perc_total = TRUE,
    direction_exc = "right",
    round_digits = 2,
    just = "center",
    just_exc = "left"
  ) |> 
  fc_split(
    N = c(1000, 800, 600), 
    label = c("Group A", "Group B", "Group C"),
    show_zero = FALSE,
    perc_total = TRUE,
    round_digits = 2,
    just = "center",
  ) |>
  fc_filter(
    N = c(800, 750, 500), 
    label = "Finished the study",
    show_exc = TRUE,
    perc_total = TRUE,
    direction_exc = "right",
    round_digits = 2,
    just = "center",
    just_exc = "left"
  ) |> 
  fc_draw(
    title = NULL,
    box_corners = "sharp"
  )
  ```
  
  Updated output:
  
![image](https://github.com/user-attachments/assets/9eadddb8-5520-4f51-ae92-eb701556688a)

kenkomodo commented:
Just a note that `prettyNum()` is a base R function - so no extra dependency is needed to make this update.

pasahe commented:
Hi Ken!

Great, it's a fantastic extension for the package, it will very usfeul to add a big mark when having big datasets. Also I love your implementation, very simple and effective. The only problem is that I think the value of the argument should be `""` by default, instead of `","`. That way we won't change the style of flowcharts created with previous versions of the package, which could sometimes be unwanted. 

Thank you so much!

pasahe reviewed (changes requested):
Please consider implementing this argument in the `fc_draw()` function instead of separately in `as_fc()`, `fc_filter()` and `fc_split()`. This way the number format will be the same in all boxes, which is desired.

Also, please change the default behavior of the argument from `","` to `""` to be consistent with the style of the existing version.

kenkomodo commented:
@pasahe I've adopted your suggestions. It required a slightly more complex implementation to make it work in `fc_draw()` instead of at each separate step, but overall still pretty simple. One thing I like about this updated approach is that it should work for reformatting `n` or `N` regardless of if the user changes the `text_pattern` argument in one of the functions before drawing the flowchart.

After update:
```
as_fc(N = 3000) |> 
  fc_filter(
    N = 2400, 
    label = "Randomized patients", 
    show_exc = TRUE,
    perc_total = TRUE,
    direction_exc = "right",
    round_digits = 2,
    just = "center",
    just_exc = "left"
  ) |> 
  fc_split(
    N = c(1000, 800, 600), 
    label = c("Group A", "Group B", "Group C"),
    show_zero = FALSE,
    perc_total = TRUE,
    round_digits = 2,
    just = "center",
  ) |>
  fc_filter(
    N = c(800, 750, 500), 
    label = "Finished the study",
    show_exc = TRUE,
    perc_total = TRUE,
    direction_exc = "right",
    round_digits = 2,
    just = "center",
    just_exc = "left"
  ) |> 
  fc_draw(
    title = NULL,
    big.mark = ",",                   # New big.mark argument
    box_corners = "sharp"
  )
  ```

  Gets:
  
![image](https://github.com/user-attachments/assets/a88d7237-38db-430e-a37d-eeebe58afbfa)

pasahe commented:
@kenkomodo it's true that specifying it in `fc_draw()` makes it a bit more complex. I've noticed that there's a problem with your current implementation because you force text to be a character when it's also possible for text to be an expression, which allows you to include bold labels and mathematical formulas. For example, the following code doesn't produce the desired output:

```
safo |> 
    as_fc(label = expression(paste("Patients ", italic("assessed"), " for ", bold("eligibility")))) |>
    fc_draw()
```
![image](https://github.com/user-attachments/assets/6d3c33a6-0924-4c15-aafc-1f69b8c890b0)

I've suggested an alternative implementation that takes this scenario into account. I've also used the `str_replace_all()` function to change all the numbers found in the box to the correct format with the appropriate big mark.

pasahe commented on R/fc_draw.R:
```diff
@@ -69,10 +70,21 @@ fc_draw.fc <- function(object, box_corners = "round", arrow_angle = 30, arrow_le
+                              # Replace any occurrence of the raw number (as character) with the formatted version.
+                              # Use fixed=TRUE so the number is treated as a literal string.
+                              text <- gsub(as.character(arg$N), N_formatted, text, fixed = TRUE)
+                              text <- gsub(as.character(arg$n), n_formatted, text, fixed = TRUE)
```
I suggest to replace this piece of code:

```
# Format n and/or N with prettyNum() using the provided big.mark
n_formatted <- prettyNum(arg$n, scientific = FALSE, big.mark = big.mark)
N_formatted <- prettyNum(arg$N, scientific = FALSE, big.mark = big.mark)
# Take the already-computed text and ensure it is a single string
text <- as.character(arg$text)
if (length(text) != 1) text <- paste(text, collapse = "\n")
# Replace any occurrence of the raw number (as character) with the formatted version.
# Use fixed=TRUE so the number is treated as a literal string.
text <- gsub(as.character(arg$N), N_formatted, text, fixed = TRUE)
text <- gsub(as.character(arg$n), n_formatted, text, fixed = TRUE)
```
for the following one:

```
#If big.mark is changed from the default:
text <- arg$text

if(big.mark != "") {
  
  if(is.character(text)) {
    
    text <- stringr::str_replace_all(text, "(\\d+)", ~prettyNum(., scientific = FALSE, big.mark = big.mark))
    
  } else {
    
    warning('The big.mark argument cannot be changed if using expressions in some boxes. It will be reset to the default (big.mark = "")')
    
  }
  
}
```

I think it's a more direct implementation using `str_replace_all()` and also takes in account the possible scenario where text is an expression giving a warning.

pasahe commented:
Great! Now it even works for expressions, that's fantastic. I just had a small question about your implementation to simplify the code a bit.

Thanks a lot @kenkomodo!

kenkomodo commented:
> Great! Now it even works for expressions, that's fantastic. I just had a small question about your implementation to simplify the code a bit.
> 
> Thanks a lot @kenkomodo!

@pasahe I'm not seeing the question you're referring to - maybe it's saved in a review that you started but didn't officially finish? If so, then it won't be visible to me until you "finish" the review.

pasahe commented:
Yes I didn't click on the button, sorry. Now you should be able to see it. Thanks!

kenkomodo commented on R/utils.R:
```diff
@@ -189,3 +189,187 @@ is_class <- function(x, class) {
+
+  # Handle language expressions (function calls)
+  if (is.call(expr)) {
+    # Special handling for 'atop' expressions
```
You're right - I think this can be simplified. I was overthinking things when I added this part.

kenkomodo commented:
@pasahe I updated things to simplify that helper function in response to your comment.

I also noticed that there was a warning that `prettyNum()` throws if the user's options are set to have `OutDec == "."`, so I added a helper function to quiet the warnings.

If the user wants to change their decimal mark in the flowchart from `"."` to `","` then they can do that in their environment without needing to add any kind of extra functionality into any of the package functions.

If the user sets `big.mark = "."` and prefers their decimal mark to be `","` instead of `"."`, then they just need to run `options(OutDec = ",")` and R will handle replacing the decimal mark:
```
# Change R enviornment options
options(OutDec = ",")

safo |>
  as_fc(label = "Patients assessed for eligibility") |>
  fc_filter(
    !is.na(group), 
    label = "Randomized", 
    show_exc = TRUE,
    perc_total = TRUE
  ) |>
  fc_draw()
```
![image](https://github.com/user-attachments/assets/5bfa468e-be3b-4cff-8a0c-2e5468ec2017)

pasahe commented:
@kenkomodo don't you think it would be good to have a warning when `OutDec` matches `big.mark`? I think the user should be warned against using the same character for both (e.g. `.` or `,``) as it can be confusing for the reader of the flowchart. So I don't see any reason to suppress the warning in `prettyNum()`. 

Sorry for all the back and forth. Let me know what you think.

kenkomodo commented:
@pasahe I originally removed it because I figured it might be obvious to the user when they look at the flowchart, but I am fine with the warning getting thrown.

Instead of removing the suppression, I added a more informative warning that alerts users when `big.mark == dec` and tells them how to change `OutDec` in their R environment options so they don't have to go searching for how to do that if they are not already familiar. Let me know what you think.

pasahe commented:
@kenkomodo perfect, I also agree that having a more informative warning will be helpful in this situation. 

I have tested it and the problem I see is that it is repeated more than once for flowchart. For example, if you run the following code, you will get this warning twice:

```
options(OutDec = ",")
as_fc(N = 9000, label = expression(paste("Patients ", italic("assessed"), " for ", bold("eligibility")))) |>
    fc_filter(N = 5000, label = expression(paste("Filter ", italic("step"), " 1 ", bold("testing")))) |>
    fc_draw(big.mark = ",")
```

_Warnings:
1: You have set big.mark equal to your environment's OutDec (',') - it can be confusing if your flowchart uses the same mark for both. Consider an alternative decimal mark.
    To change the decimal mark, run: `options(OutDec = "<alternative decimal mark>")` 
2: You have set big.mark equal to your environment's OutDec (',') - it can be confusing if your flowchart uses the same mark for both. Consider an alternative decimal mark.
    To change the decimal mark, run: `options(OutDec = "<alternative decimal mark>")`_ 

Perhaps you could suppress the warning in `quiet_prettyNum()`, which is run multiple times, and put your custom warning in `update_numbers()`, which is run only once. With this last change, the pull request is ready to merge.

pasahe commented on R/utils.R:
```diff
@@ -189,3 +189,155 @@ is_class <- function(x, class) {
+
+#' @title quiet_prettyNum
+#' @description Wrapper for `prettyNum()` that quiets warning message if user sets `big.mark = "."`.
+#'
```
In the description I would say that the function silences the warning not only when user sets `big.mark = "."`, but also when user sets the same big.mark as the character defined in the OutDec option.

pasahe commented on R/utils.R:
```diff
@@ -189,3 +189,155 @@ is_class <- function(x, class) {
+  #    "'big.mark' and 'decimal.mark' are both '.', which could be confusing"
+  if (big.mark == dec) {
+    warning(sprintf("You have set big.mark equal to your environment's OutDec ('%s') - it can be confusing if your flowchart uses the same mark for both. Consider an alternative decimal mark.\n    To change the decimal mark, run: `options(OutDec = \"<alternative decimal mark>\")`", dec),
+            call. = FALSE)
```
I would move this warning message from this section to the beginning of `update_numbers()`

kenkomodo commented:
@pasahe Great suggestion! I am traveling, so I may not get around to adopting this last tweak until later this week or next week. But I will make the update when I get the chance.

pasahe commented:
@kenkomodo No worries! Take your time, there is no rush.

Thanks a lot!!

kenkomodo commented:
@pasahe I tried moving the warning as you suggested, and that same piece of code ended up throwing 26 warnings instead of 2! 😆 

I updated the helper function so that it uses `rlang::warn()` instead and will only throw the error once every 8 hours for each unique `dec` when `big.mark == dec`.

Now when I run that same code you listed, I get this output:
```
> options(OutDec = ",")
> as_fc(N = 9000, label = expression(paste("Patients ", italic("assessed"), " for ", bold("eligibility")))) |>
+     fc_filter(N = 5000, label = expression(paste("Filter ", italic("step"), " 1 ", bold("testing")))) |>
+     fc_draw(big.mark = ",")
Warning message:
ℹ You have set big.mark equal to your environment's OutDec (',') - it can be confusing if your flowchart uses the same mark for both. Consider an alternative decimal mark.
→ To change the decimal mark, run: `options(OutDec = "<alternative decimal mark>")`
This warning is displayed once every 8 hours. 
```

Running the same code in the same R session yields no additional warning messages.

I also updated the function name and documentation to align it with the current version of the function as you requested.

I think this should cover the PR, but let me know if I missed something.

pasahe commented:
Hi @kenkomodo!

26 warnings is too many warnings, even for a user who sets the same big mark as decimal number😆.

It's weird because I tried to put a warning at the beginning of the `update_numbers()` function and it only shows up once. I will show you in the review where I put the warning to see if it fixes your problem. If not, the current workaround works fine.

Thanks!!

pasahe commented on R/utils.R:
```diff
@@ -189,3 +189,164 @@ is_class <- function(x, class) {
+update_numbers <- function(object, big.mark = "") {
+  # Handle both tibble and list formats
+  fc_list <- if(tibble::is_tibble(object$fc)) list(object$fc) else object$fc
+
```
If you place the following code here and remove it from the `info_prettyNum()` function, the warning should only appear once (at least it does for me):

```
# Get the current decimal mark from user environment, if not set then return "."
  dec <- getOption("OutDec", ".")

  # if user specifies `big.mark = "."` and `dec == "."`, then suppress the warning that will pop up:
  #    "'big.mark' and 'decimal.mark' are both '.', which could be confusing"
  if (big.mark == dec) {
    message <- c(
      "i" = sprintf("You have set big.mark equal to your environment's OutDec ('%s') - it can be confusing if your flowchart uses the same mark for both. Consider an alternative decimal mark.", dec),
      ">" = "To change the decimal mark, run: `options(OutDec = \"<alternative decimal mark>\")`"
    )
    warning(message)
  }
```

kenkomodo commented:
@pasahe Fixed without 26 warnings this time! I think I may have moved the warning to the wrong helper function instead of `update_numbers()` and didn't notice because it was before my coffee kicked in, which is probably what led to the 26 warnings! 😆 

In any event, I made the update and reverted the name from `info_prettyNum()` back to `quiet_prettyNum()`. I also left the warning call using `rlang::warn()` instead of `warning()` since `rlang` is already a dependency and it allows the warning to be formatted. I removed the arguments that limited the waring to once every 8 hours though.

pasahe commented:
Great! Now everything works perfectly. Sorry for bothering you with so many minor changes...

Thanks a lot!! It's a great update for the package 🚀

kenkomodo commented:
No worries! Thanks or humoring my suggestions for updates!
